Hj docs updates

hnjolles1 opened the pull request:
### Overview
_Provide a brief description of what this PR does, and why it's needed._

This PR started with updating DEVNET.md and CONTRACTS.md with the 0.9 addition of TTL and Deploy Dependences. Additionally, this PR now includes updates to formatting and code examples as well as updates to content not included from prior sprints. See list of JIRA tickets below
### Which JIRA ticket(s) does this PR relate to?
_Add the link here Create a ticket and link it here if one does not exist.
_[ECO-181 ](https://casperlabs.atlassian.net/browse/ECO-181) 
_[ECO-223](https://casperlabs.atlassian.net/browse/ECO-223)
_[ECO-267](https://casperlabs.atlassian.net/browse/ECO-267)
_[ECO-291](https://casperlabs.atlassian.net/browse/ECO-291)

### Complete this checklist before you submit this PR
- [ ] This PR contains no more than 200 lines of code, excluding test code.
- [ ] This PR meets [CasperLabs coding standards](https://casperlabs.atlassian.net/wiki/spaces/EN/pages/16842753/Coding+Standards).
- [ ] If this PR adds a new feature, it includes tests related to this feature.
- [x] You assigned one person to review this PR.
- [x] Your GitHub account is linked with our [Drone CI](https://drone-auto.casperlabs.io/) system. This is necessary to run tests on this PR.
- [ ] Do not forget to run `bors r+` if GitHub policy is not enforced, e.g. when merging into another feature branch. It may be omitted under some circumstances if this PR intentionally assumes that integration tests will fail but will be fixed with the future PRs.

### Notes
_Optional. Add any notes on caveats, approaches you tried that didn't work, or anything else._
When pushing updates with addition

hnjolles1 commented:
Hi please see changes requested from PR 1516 and review this documentation re: additions for Time to Live and Deploy Dependency.

piokuc commented on docs/CONTRACTS.md:
```diff
@@ -225,58 +237,47 @@ Note: names are valid only in the context of the account which called `add_uref`
-by calling Contract API function `get_arg`
-with index of an argument.
-First argument is indexed with `0`.
+**Time to Live** specifies a duration by which the deploy can be included in a block prior to expiration. The node will not accept deploys with deploy.timestamp greater than some configurable number of milliseconds in the future (relative to its current time).  This maximum future time is configurable because this parameter is not visible at the protocol level since deploys can only go into block ...
```
I find it hard to understand this sentence: "This maximum future time is configurable because this parameter is not visible at the protocol level since deploys can only go into blocks after their deploy.timestamp."  Perhaps it should be split into shorter, more clear sentences.

piokuc commented on docs/CONTRACTS.md:
```diff
@@ -225,58 +237,47 @@ Note: names are valid only in the context of the account which called `add_uref`
-and its JSON serialized form is ready for this change.
+Note, contract arguments are positional,
+and so the `"name"` attribute is currently not used. However, we plan to change contract arguments to be keyword (named) arguments.
+The structure of the `Arg` protobuf message and its JSON serialized form is ready for this change.
```
This information is obsolete, because the change anticipated in it has been already implemented. Therefore it can be removed entirely.

piokuc replied:
Because of the above mentioned change section "Accessing arguments in contracts" is no longer correct. Please update examples in that section that show accessing contract arguments via their indices (`get_args(0)`). Currently `get_arg` accepts name of a parameter as a string, for example: `get_arg("amount")`. 

Also: "will deserialize first contract argument as a value of type u64" should be changed to "will deserialize contract argument `amount` as a value of type u64"

piokuc commented on docs/CONTRACTS.md:
```diff
@@ -225,58 +237,47 @@ Note: names are valid only in the context of the account which called `add_uref`
-to accept a string with a name of an argument
-instead of it's index.
+In a future release Contract API `get_arg` function
+will change to accept a string with a name of an argument instead of it's index.
```
The same as above: this information is now obsolete, please remove.

piokuc commented on docs/DEVNET.md:
```diff
@@ -1,24 +1,31 @@
-* The request will appear in the "Recent Faucet Requests" table; wait for the green check mark to appear in the "Status" column
+ The request will appear in the "Recent Faucet Requests" table; wait for the green check mark to appear in the "Status" column 
+ 
+  **Note:** that a "token" obtained on devnet has no monetary value whatsoever
```
A dot ending the sentence is missing here and in the sentence above.

piokuc commented on docs/DEVNET.md:
````diff
@@ -38,14 +45,58 @@ casperlabs-client \
+casperlabs-client\ 
+        --host deploy.casperlabs.io \ 
+        --port 40401 show-deploy <deploy-hash>
+```
````
The content of the section on bonding seems to be a copy of the text about monitoring deploy status.

hnjolles1 replied:
Hi Peter, I'm not sure what file you are looking at, this is nothing like the one I'm working on.

piokuc replied:
Sorry, Hope, I got confused by the diff. "You can also monitor the outcome of deploys ..." appears in the diff under line 45 "### Bonding", but this is a deleted line... I think I went color blind for a moment and didn't notice the pinkish color that signifies deletion. Sorry about that.

hnjolles1 replied:
That's life.

piokuc commented on docs/DEVNET.md:
```diff
@@ -38,14 +45,58 @@ casperlabs-client \
+
+* [Install the CasperLabs client](INSTALL.md)
+* [Compile a contract written in rust](CONTRACTS.md)
+* Use the CasperLabs client `deploy` sub-command (example below)
```
The items "Install the CasperLabs client", "Compile a contract written in rust" and "Use the CasperLabs client `deploy` sub-command" are already listed in the Step 3. I don't see a point of repeating them in Step 4.

hnjolles1 commented on docs/CONTRACTS.md:
```diff
@@ -82,13 +89,17 @@ You should see the following output:
+
+The `deploy` command does it all (creates a deploy, signs it, deploys to the node, and allows for monitoring the outcome of deploys) but doesn't allow for signing with multiple keys. 
 
-Every account can associate multiple keys with it and give each a weight. Collective weight of signing keys decides whether an action of certain type can be made. In order to collect weight of different associated keys a deploy has to be signed by corresponding private keys. `deploy` command does it all (creates a deploy, signs it and deploys to the node) but doesn't allow for signing with multip ...
```
re: 87 . this is odd, it is not in the file i pushed

hnjolles1 replied:
87 -- this is odd, this is not in the file i uploaded, in any case, will make sure to update as per your comment

birchmd commented on docs/CONTRACTS.md:
```diff
@@ -225,84 +241,73 @@ Note: names are valid only in the context of the account which called `add_uref`
-but find it missing in the context that they are actually run in, that is of the deployer's account.
+In contrast, when the same stored contract is called directly.  For example, its address is passed to `--session-hash` argument of the `deploy` command, the contract will be executed in the context of the account that creates the deploy. 
+
+The consequence of this is that stateful contracts designed to operate in a specific context may not work as expected when called directly. They may, for instance, attempt to read or modify a `URef` that they expect to exist in their context, but find it missing in the context that they are actually run in, that is of the deployer's account.
```
We should mention the way to get around this (for now anyways, this will likely change in the future) is to create a stateless "proxy" contract which essentially does `call_contract` and nothing else. We can point to an example of a stored contract which does this; I have a PR open for one now, but it's not merged yet: https://github.com/CasperLabs/CasperLabs/pull/1527

birchmd commented on docs/CONTRACTS.md:
```diff
@@ -225,84 +241,73 @@ Note: names are valid only in the context of the account which called `add_uref`
+
+**Time to Live and Deploy Dependency**
+
+**Time to Live** specifies a duration for which the deploy can be included in a block prior to expiration. The node will not accept deploys with deploy.timestamp greater than some configurable number of milliseconds in the future (relative to its current time). This maximum future time is configurable -- this parameter is not at the protocol level since deploys can only go into blocks after their ...
```
`deploy.timestamp` should be formatted as code.
It is important to point out that individual nodes choose their own future cut-off, so this is something set by node operators, not users.

birchmd commented on docs/CONTRACTS.md:
```diff
@@ -225,84 +241,73 @@ Note: names are valid only in the context of the account which called `add_uref`
 
-Contract API function `get_arg` allows to access contract arguments,
-for example: 
+Contract API function `get_arg` allows to access contract arguments. accessing contract arguments via their indices `(get_args(0))`. Currently `get_arg` accepts name of a parameter as a string.
```
This is not true, arguments cannot be accessed by name as of now; it is by index only.

birchmd commented on docs/DEVNET.md:
```diff
@@ -1,24 +1,35 @@
-* The request will appear in the "Recent Faucet Requests" table; wait for the green check mark to appear in the "Status" column
 
-#### Checking the balance of an account
+  Provide a name for your account and click [Save]; multiple key files will be written to disk. You will need these key files to interact with the network, so give some consideration to where you save them.
```
Leading spaces?

birchmd commented on docs/DEVNET.md:
````diff
@@ -27,26 +38,74 @@ casperlabs-client \
 ```
 
-### Bonding
+* You can also monitor the outcome of deploys using the casperlabs-client:
````
Again, I don't like the connotation of "monitor", this is a single poll / query of the status.

birchmd commented on docs/DEVNET.md:
```diff
@@ -27,26 +38,74 @@ casperlabs-client \
+
+* To add Parameters for Time to Live and Deploy Dependencies:
+
+* The `time_to_live` parameter is included in the deploy to set a time frame limiting how long a deploy is effective before it is no longer valid.
```
Seems like some duplicate information here, maybe we should just link to `CONTRACTS.md`?

piokuc commented:
@hnjolles1: please note, it is the reviewer who is supposed to resolve conversations, once (s)he is satisfied with changes done by the PR author.

hnjolles1 commented on docs/CONTRACTS.md:
```diff
@@ -147,62 +157,68 @@ casperlabs-client sign-deploy --private-key [private_key] --public-key [public_k
-The function is stored on blockchain with `store_function`.
-Next, a call to `add_uref` associates the stored contract's address with a name `"counter"`.
+In the example below `counter_ext` is a function in the same module as the executing contract.
+The function is stored on blockchain with `store_function`. Next, a call to `add_uref` associates the stored contract's address with a name `"counter"`.
```
@birchmd just double checking that throughout CL  'add_uref' no longer exists and is now 'put_key', and not just `put_key` in this instance

birchmd replied:
Yes, that is correct.

hnjolles1 commented on docs/CONTRACTS.md:
```diff
@@ -147,62 +157,68 @@ casperlabs-client sign-deploy --private-key [private_key] --public-key [public_k
 
+**Monitoring deploy status**
+
+The `casperlabs-client--help` says that `show-deploy` allows the user to `View properties of a deploy known by Casper on an existing running node`. One of the properties of a deploy is its status which can be one of:
```
@birchmd  promise to update docs, etc. to make sure to never make readers feel like "plebs" ever

birchmd commented on docs/CONTRACTS.md:
```diff
@@ -164,89 +165,100 @@ For more detailed description about deploy commands, use the `--help` flag (`cas
-Typically this is done in the same contract that calls `store_function`.
-In the example below `counter_ext` is a function in the same module as the executing contract.
-The function is stored on blockchain with `store_function`. Next, a call to `add_uref` associates the stored contract's address with a name `"counter"`.
+In the example below, `counter_ext` is a function in the same module as the executing contract. The function is stored on blockchain with `store_function`. Next, a call to `put_key` associates the stored contract's address with a name `"counter"`.
```
Since I have updated the `counter` contract in dev, we could use it as an instead of this custom one. As is the example would need updating because it still contains usages of `add_uref`.

birchmd commented on docs/CONTRACTS.md:
```diff
@@ -164,89 +165,100 @@ For more detailed description about deploy commands, use the `--help` flag (`cas
+Note that for now, one way to get around this (as this will likely change in the future), is to create a stateless "proxy" contract which essentially does `call_contract` and nothing else.
 
-The consequence of this is that stateful contracts designed to operate in a specific context may not work as expected when called directly. They may, for instance, attempt to read or modify a `URef` that they expect to exist in their context, but find it missing in the context that they are actually run in, that is of the deployer's account.
+For example, this stored contract: Increment the counter
```
Seems like `this` should be a link.

birchmd commented on docs/CONTRACTS.md:
````diff
@@ -164,89 +165,100 @@ For more detailed description about deploy commands, use the `--help` flag (`cas
+    --payment-amount 10000000 \
+    --session-name counter_inc
+```
+The counter contract example, stores an additional function for incrementing the counter. This function can be called directly by name in order to increment the counter without sending additional wasm. This is useful because it means we can greatly reduce the amount of wasm we send in the current LRT workflow.
````
> This is useful because it means we can greatly reduce the amount of wasm we send in the current LRT workflow.

We don't need to say this in our public documentation.

birchmd commented on docs/CONTRACTS.md:
```diff
@@ -130,179 +141,213 @@ casperlabs-client \
-acting as the deploy's session contract.
 
-Equivalent argument for payment contract is `--payment-name`.
+**Calling a stored contract by reference**
```
This is not a new example compared to what is already here. But the fact that you thought it was perhaps means we need to work on this section some more to make it clearer.

There are two ways to store a contract under a `Key`: `store_function` (stores under a `Key::URef`) and `store_function_at_hash` (stores under a `Key::Hash`).

There is one way to associate a `Key` with a human-readable name: `put_key`. This association is only valid in the context where `put_key` is run.

From the client (either scala or python) there are 4 ways to execute code:
- Provide the wasm directly (this corresponds to `--session` or `--payment`, depending on what execution phase you want the code to be used for)
- Provide the `Key::Hash` address in base-16 that a contract was previously stored under via `store_function_at_hash` (this corresponds to `--session-hash` or `--payment-hash`)
- Provide the human readable name that is associated with some `Key` (via `put_key`) that a contract was previously stored under via either `store_*` function (this corresponds to `--session-name` or `--payment-name`)
- Provide the bytes identifying the `Key::URef` that a contract was previously stored under (this corresponds to `--session-uref` or `--payment-uref`)

The last one I don't even think we should document because in order to use it you must have already called `put_key` to persist the `URef` in the account (otherwise you will get a forged `URef` error), which means you should just use the human readable name anyways.

So, based on the cases above, we need two rust source examples, one for `store_function` and one for `store_function_at_hash` (these could really be one example, but I don't necessarily want to confuse the matter). We also need two client CLI examples, one for `--session-hash` (which only works with contracts stored at hashes of course) and one for `--session-name` (which works with either storage function); the one with passing in the wasm itself rather than referencing a previously stored contract is already covered in the basic description of deploy.

It might also be worth calling out that the workflow here would be store store a contract by sending the wasm (using `--session) then calling the contract by name/address (using `--session-name` / `--session-hash`). This is sort of laid out in the [count-define example](https://github.com/CasperLabs/CasperLabs/tree/dev/execution-engine/contracts/examples/counter-define), via the two client CLI snippets in the README, but not at all explained, so it might be nice to do that in this document.

If all that is clear then please refactor the whole `Stored contracts` section to reflect all this in the best way you can.

hnjolles1 commented:
> Thanks for making those changes. I noticed a couple more things, but I think we're just about there.

The changes were done, updated the custom example and added a second example as best as possible but was not confident i had completely understood it.
> 
> Also, I don't think you meant to delete `Casper.scala` and just had a bit of trouble with the merge with dev. After we sort out the doc changes we want to do maybe we will just rebase this branch to fix the problem.

am mortified, i definitely did not mean to, delete the `Casper.scala` -- this was a mistake from panic that that should not have been with the PR- did not want to mess up a more recent version in dev in the merge. 

Thanks for understanding and yes let's rebase.

piokuc commented on docs/CONTRACTS.md:
````diff
@@ -130,179 +141,213 @@ casperlabs-client \
 
+For example:
+```
+get_arg("amount")
````
Sorry, I was wrong about this, named arguments haven't been implemented yet. Please remove lines 330-332, example on line 328 is enough.

ashok-cl commented:
@hnjolles1 and @piotr-dziubecki - where are we on this PR? Should we wait for it to get merged? It has been open for almost a month now.

birchmd commented on docs/CONTRACTS.md:
```diff
@@ -143,15 +143,15 @@ casperlabs-client \
 In the example above there is no `-i` argument, meaning that the signed deploy will be read from STDIN.
 
-Reading from STDIN and writing to STDOUT allows for piping output from one command to the input of another one (commands are incomplete for better readability):
+Reading from STDIN and writing to STDOUT allows for piping output from one command to the input of another one ( commands are incomplete for better readability):
```
Extra space after left parenthesis

birchmd commented on docs/CONTRACTS.md:
```diff
@@ -191,131 +191,150 @@ You can also retrieve further information from our platform (with our APIs, et.
-A function that is part of the deployed contract's module can be saved on the blockchain with the Contract API function `store_function`.
-
-Such function becomes a stored contract that can be later called from another contract with `call_contract`, or used instead of a WASM file when creating a new deploy on the command line.
+A `store_function` that is part of the deployed contract's module can be saved on the blockchain with the Contract API `store_function`. This function becomes a stored contract that can later be called from another contract with `call_contract`, or used instead of a WASM file when creating a new deploy on the command line.
```
```suggestion
A function that is part of a session code module can be saved on the blockchain with the Contract API `store_function`. This function becomes a stored contract that can later be called from another contract with `call_contract`, or used as payment/session code, instead of a wasm file, when creating a new deploy.
```

birchmd commented on docs/CONTRACTS.md:
```diff
@@ -191,131 +191,150 @@ You can also retrieve further information from our platform (with our APIs, et.
 
-- `store_function` stores a contract under a `Key::URef`
-- `store_function_at_hash` stores a contract under a `Key::Hash`
+There are two ways to store a contract under a Key: `store_function` stores a contract under a `Key::URef`, `store_function_at_hash` stores a contract under a `Key::Hash` -- a 256-bit unique identifier which is generated by the system. Both correspond to `--session-name` or `--payment-name`.
```
No, this is not correct. `--session-name` and `--payment-name` will reference a named key in the deployer's account. If you call `store_function` then call `put_key` to store the key under a name then that name will correspond to `--session-name` and/or `--payment-name` (similarly for `store_function_at_hash`).

birchmd commented on docs/CONTRACTS.md:
```diff
@@ -191,131 +191,150 @@ You can also retrieve further information from our platform (with our APIs, et.
-The following are ways to execute code from the client ~~(e.g., scala or python)~~:
-
-- Provide the wasm directly, this corresponds to `--session` or `--payment` depending on what execution phase you want the code to be used for.
+For Example: `store_function`
```
Unnecessary capital 'E'

birchmd commented on docs/CONTRACTS.md:
````diff
@@ -191,131 +191,150 @@ You can also retrieve further information from our platform (with our APIs, et.
-- Provide the wasm directly, this corresponds to `--session` or `--payment` depending on what execution phase you want the code to be used for.
+For Example: `store_function`
+ 
+``` 
````
You can make code blocks have syntax highlighting by putting the language name after the triple ticks.

Also, the last line in this code block does not look like valid rust syntax. And do we ever explain what `counter_urefs` is all about?

birchmd commented on docs/CONTRACTS.md:
```diff
@@ -191,131 +191,150 @@ You can also retrieve further information from our platform (with our APIs, et.
+**Associating a Key with a Name**
 
-`    fpo                 `
+There is one way to associate a Key with a human-readable name with `put_key`. This association (names) is valid only in the context where `put_key` is run.
```
This is a true sentence, but I wonder how meaningful it is. Do we explain earlier on what a "context" is? There also doesn't appear to be any motivation for this section. Why would I want to associate a key with a human readable name in the first place? (There are good reasons of course, I just don't know if we are describing them explicitly).

birchmd commented on docs/CONTRACTS.md:
```diff
@@ -191,131 +191,150 @@ You can also retrieve further information from our platform (with our APIs, et.
+**Calling stored contracts**
 
-For example:
+From the client, there are several ways you can execute code: as a direct deploy (see basic description of [deploy](https://github.com/CasperLabs/CasperLabs/blob/release-v0.11/docs/CONTRACTS.md#step-5-deploy-counterdefinewasm)), or a stored contract whereby payment and session contracts stored on the chain are called providing either a `hash`, `name`, or `uref` of the contract.
```
I think you can make the URL just `CONTRACTS.md#step-5-deploy-counterdefinewasm` and it will always link to the current version of this document; instead of us needed to try to remember to update the release version all the time.

Also, what is the distinction you are drawing between a "direct deploy" vs a deploy where the payment/session code are stored on-chain already? I have been using the terms interchangeably.

birchmd commented on docs/CONTRACTS.md:
```diff
@@ -191,131 +191,150 @@ You can also retrieve further information from our platform (with our APIs, et.
+From the client, there are several ways you can execute code: as a direct deploy (see basic description of [deploy](https://github.com/CasperLabs/CasperLabs/blob/release-v0.11/docs/CONTRACTS.md#step-5-deploy-counterdefinewasm)), or a stored contract whereby payment and session contracts stored on the chain are called providing either a `hash`, `name`, or `uref` of the contract.
 
-`    fpo   `
+You can call an existing contract stored on the blockchain using its address, or sending the arguments to the contract. The deploy command accepts a list of arguments for either (or both) of the payment and session contracts, as values of parameters `--session-args` and `--payment-args`.
```
Passing arguments is a totally independent feature from stored contracts. You can also pass arguments to contracts when you give them as explicit wasm code as well. We should have a distinct section all about passing arguments if we do not have one already.

birchmd commented on docs/CONTRACTS.md:
````diff
@@ -191,131 +191,150 @@ You can also retrieve further information from our platform (with our APIs, et.
+```
 
-Note that if you don't use the human readable name, you will get a forged `URef error`. You must have first called the `put_key` to persist the `URef` in the account, in which case you would have used the key initially.
+Note that `Standard Payment Code` is used as the default where no payment code is supplied for the deploy.
````
This note feels out of place here. Seems like it belongs in a description about the client, rather than about contracts in general.

birchmd commented on docs/CONTRACTS.md:
```diff
@@ -191,131 +191,150 @@ You can also retrieve further information from our platform (with our APIs, et.
-The `store_function` stores under a `URef`, while `store_function_at_hash` stores under a `Hash` -- a 256-bit unique identifier which is generated by the system.
-
-It is important to note that when a contract is stored under a `Hash` it is immutable (that `Hash` will always point to exactly that contract), while storing under a `URef` allows the contract to be upgraded with the `upgrade_contract_at_uref` function.
+Contract address is a cryptographic hash uniquely identifying a stored contract in the system. Thus, it can be used to call the stored contract directly when creating a deploy, e.g. on command line; or from another contract.
```
Calling a contract via `call_contract` does not create a new deploy. So this sentence is almost true, but needs to be reworked a little so that it does not imply calling from another contract is creating a deploy.

birchmd commented on docs/CONTRACTS.md:
````diff
@@ -191,131 +191,150 @@ You can also retrieve further information from our platform (with our APIs, et.
-`--session-hash 2358448f76c8b3a9e263571007998791a815e954c3c3db2da830a294ea7cba65`.
+For example:  Calling a stored contract by `--session-hash`
+```
+$ casperlabs-client --host $HOST deploy \     
````
Random character at the end of some lines.

birchmd commented on docs/CONTRACTS.md:
```diff
@@ -191,131 +191,150 @@ You can also retrieve further information from our platform (with our APIs, et.
 **Calling a stored contract by name**
 
-A contract address can be associated with a name in the context of a user's account. Typically this is done in the same contract that calls `store_function`.
+For convenience, a contract address can be associated with a human readable name in the context of a user's account.
```
Didn't we have a section about this already above?

MParlikar commented:
What is the status of this PR @hnjolles1 ?  It's been ~ 2 weeks without movement.  Should this PR be closed?

hnjolles1 commented:
restructured CONTRACTS.md -- see upcoming PRs separate for CONTRACTS.md and DEVNET.md
